### PR TITLE
[MAINTENANCE] Update remaining atomic prescriptive templates (1 of 2)

### PR DESCRIPTION
**Pull request by NathanFarmer**
Changes proposed in this pull request:
- GREAT-1478
- Update about a dozen remaining `_atomic_prescriptive_template` methods to best practice
- Fixes 7 latent renderer bugs and 2 incorrect test cases

### Definition of Done
- [X] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [X] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [X] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

[GREAT-1478]: https://greatexpectations.atlassian.net/browse/GREAT-1478?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py
Comment (NathanFarmer) [inline tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py:659]: This was caused by an incorrect test case and should not have been treated as a correct renderer output.

File: tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py
Comment (NathanFarmer) [inline tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py:904]: These were missing due to a bug. You will notice they are referenced in the `template` string on line 919, but the param substitutions are missing from the payload.

File: tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py
Comment (NathanFarmer) [inline tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py:995]: Another bug. `$column_list_0` would not have been substituted by the `v__0` param.

File: tests/expectations/test_expectation_atomic_renderers.py
Comment (NathanFarmer) [inline tests/expectations/test_expectation_atomic_renderers.py:725]: This Expectation kwarg does not accept integers. See "Keyword Args" for this Expectation [here](https://greatexpectations.io/expectations/expect_column_values_to_be_increasing?filterType=Backend%20support&gotoPage=1&showFilters=false&viewType=Summary).

File: tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py
Comment (NathanFarmer) [inline tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py:592]: This was caused by an incorrect test case and should not have been treated as a correct renderer output.

File: tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py
Comment (NathanFarmer) [inline tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py:754]: This was a bug. This `formatted_json` was not being rendered previously.

File: tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py
Comment (NathanFarmer) [inline tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py:926]: Another bug. `$column_list_0` would not have been substituted by the `v__0` param.

File: tests/expectations/test_expectation_atomic_renderers.py
Comment (NathanFarmer) [inline tests/expectations/test_expectation_atomic_renderers.py:667]: This Expectation kwarg does not accept integers. See "Keyword Args" for this Expectation [here](https://greatexpectations.io/expectations/expect_column_values_to_be_increasing?filterType=Backend%20support&gotoPage=1&showFilters=false&viewType=Summary).

File: tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py
Comment (NathanFarmer) [inline tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py:909]: The commas are removed here so that we can get consistent rendering behavior for arrays (other renderers don't add the commas). Styling is currently applied to the `$` params, so differentiating between params is easy.

File: tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py
Comment (NathanFarmer) [inline tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py:951]: Another bug. `$column_list_0` would not have been substituted by the `v__0` param.

File: tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py
Comment (NathanFarmer) [inline tests/expectations/snapshots/snap_test_expectation_atomic_renderers.py:1010]: Another bug. `column_list` params are missing. Also, there is no reason to change the kwarg name from `column_set` to `column_list` in the renderer. This is not a breaking change since the substitution logic is still the same.

File: great_expectations/expectations/core/expect_column_distinct_values_to_contain_set.py
Comment (NathanFarmer) [inline great_expectations/expectations/core/expect_column_distinct_values_to_contain_set.py:141]: Decided to move away from this pattern having a different class method for each array-type param. There 4 (so far) different versions of `array_param_name` and `param_prefix` used throughout renderers. No, it does not make sense to make them an `Enum` since these can be anything the contributor defines.

File: great_expectations/render/renderer_configuration.py
Comment (Kilo59) [inline great_expectations/render/renderer_configuration.py:108]: `non-blocking`
Would a named tuple be helpful for this inner tuple?

File: great_expectations/expectations/expectation.py
Comment (Kilo59) [inline great_expectations/expectations/expectation.py:1611]: `non-blocking`
`RendererConfiguration` isn't a pydantic model is it? I assume no, but just double-checking.
If so you could use a `constr` with `min_length=1` to perform this same check.
https://docs.pydantic.dev/usage/types/#arguments-to-constr

File: great_expectations/expectations/core/expect_column_values_to_be_null.py
Comment (Kilo59) [inline great_expectations/expectations/core/expect_column_values_to_be_null.py:107]: ~~is `AddParamArgs` already a container/sequence type?~~

Edit: Yes

**Review by Kilo59 — APPROVED** [review]
Sorry for the late review.

A couple of non-blocking comments/questions.

Are there any of these expectations that we are type-checking?
If not would you pick one with a low number of errors and bring it under coverage to act as a kind of "canary in the coal mine"? If there's already one under coverage then nevermind.

File: great_expectations/render/renderer_configuration.py
Comment (NathanFarmer) [reply great_expectations/render/renderer_configuration.py:108]: I don't see why not. ✅

Comment (Kilo59) [reply great_expectations/render/renderer_configuration.py:108]: Prefer `typing.NamedTuple` over `collections.name_tuple`.

File: great_expectations/expectations/expectation.py
Comment (NathanFarmer) [reply great_expectations/expectations/expectation.py:1611]: `RendererConfiguration` is a pydantic model, yes. I added `constr(min_length=1)` to the `name` variable of the `add_param` method, but a check for this prefix is still required. Notice it gets prepended to a valid string below.

File: great_expectations/render/renderer_configuration.py
Comment (Kilo59) [inline great_expectations/render/renderer_configuration.py]: Is this getting dynamically added to a `pydantic` model?
I don't know if it does anything at runtime as a function arg annotation.

Comment (NathanFarmer) [reply great_expectations/render/renderer_configuration.py]: Yes this adds an attribute to the `params` on `RendererConfiguration`. Yeah I think you are correct that this won't work. `name` already can't be an empty string because it becomes the name of a `BaseModel`. I'm going to leave this as-is.

File: great_expectations/render/renderer_configuration.py
Comment (NathanFarmer) [reply great_expectations/render/renderer_configuration.py:108]: That's what I used!

File: great_expectations/expectations/expectation.py
Comment (NathanFarmer) [reply great_expectations/expectations/expectation.py:1611]: The `constr` change was reverted since `add_param` is a method on the `RendererConfiguration`.

File: great_expectations/render/renderer_configuration.py
Comment (Kilo59) [reply great_expectations/render/renderer_configuration.py]:
```
great_expectations/render/renderer_configuration.py:501: error: Invalid type
comment or annotation  [valid-type]
            name: constr(min_length=1),
                  ^
great_expectations/render/renderer_configuration.py:501: note: Suggestion: use constr[...] instead of constr(...)
```

type-checker doesn't like this. I'm not sure it really makes sense outside of `pydantic.BaseModel` field annotation.

You *can* enforce method/function calls the same way you do with init args, but it doesn't happen automatically.
https://docs.pydantic.dev/usage/validation_decorator/

**Comment by NathanFarmer** [comment]
> If not would you pick one with a low number of errors and bring it under coverage to act as a kind of "canary in the coal mine"? If there's already one under coverage then nevermind.

@Kilo59 I am updating `pyproject.toml` in this PR (will merge in develop and reduce diffs after merging this PR): https://github.com/great-expectations/great_expectations/pull/6724

There are already a few Expectations under type checking.

**Pull request merged**
